src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+
+            var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
+
+            if (document == null)
+            {
+                var listWithSelections = GetHighlightedList(initialListOfItemsToBeIncluded, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, data.SelectedFilters, EditorCompletion.UpdateSelectionHint.NoChange, centerSelection: true, uniqueItem: default);
+            }
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                var completionService = document.GetLanguageService<CompletionService>();
+                var completionRules = completionService.GetRules();
+                return HandleAllItemsFilteredOut(data.Trigger.Reason, data.SelectedFilters, selectedFilters, completionRules);
+            }
+
+            // If this was deletion, then we control the entire behavior of deletion ourselves.
+            if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion || data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Backspace)
+            {
+                return HandleDeletionTrigger(data.InitialSortedList, data.Trigger.Reason, data.SelectedFilters, data.Trigger.Reason, filterText, initialListOfItemsToBeIncluded);
+            }
+
+            var caretPoint = session.TextView.GetCaretPoint(data.Snapshot.TextBuffer);
+            var caretPosition = caretPoint?.Position;
+
+            return HandleNormalFiltering(
+                data.InitialSortedList,
+                snapshotForDocument,
+                document,
+                caretPosition,
+                filterText,
+                data.SelectedFilters,
+                data.Trigger.Reason,
+                initialListOfItemsToBeIncluded,
+                mustSetSelection);
+        }
+
+        private EditorCompletion.FilteredCompletionModel HandleDeletionTrigger(
+            ImmutableArray<EditorCompletion.CompletionItem> sortedList,
+            EditorCompletion.CompletionTriggerReason triggerReason,
+            ImmutableArray<EditorCompletion.CompletionFilterWithState> filters,
+            EditorCompletion.CompletionTriggerReason filterReason,
+            string filterText,
+            List<FilterResult> filterResults)
+        {
+            FilterResult? bestFilterResult = null;
+            int matchCount = 0;
+            foreach (var currentFilterResult in filterResults.Where(r => r.MatchedFilterText))
+            {
+                if (bestFilterResult == null ||
+                    IsBetterDeletionMatch(currentFilterResult, bestFilterResult.Value))
+                {
+                    // We had no best result yet, so this is now our best result.
+                    bestFilterResult = currentFilterResult;
+                    matchCount++;
+                }
+            }
```

Comment:
there is code missing that used to be above this.  where did it go?

Reply:
There was code assuming that there were two triggers: filterReason and model.Trigger. Now there is only one of them. Therefore, the check before calling the method already checks if it is deletion and not insertion.

---
In reply to: [222086760](https://github.com/dotnet/roslyn/pull/29016#discussion_r222086760) [](ancestors = 222086760)